File: 2185-counting-words-with-a-given-prefix/2185-counting-words-with-a-given-prefix.py
class TrieNode: 
    def __init__(self):
        self.children = {}
        self.count = 0

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def add(self, words: List[str]):
        for word in words:
            curr = self.root
            for c in word:
                if c not in curr.children:
                    curr.children[c] = TrieNode()
                curr = curr.children[c]
                curr.count += 1

    def prefixCount(self, pref: str) -> int:
        curr = self.root
        for p in pref: 
            if p not in curr.children: return 0
            curr = curr.children[p]
        return curr.count

class Solution:
    def prefixCount(self, words: List[str], pref: str) -> int:
        res = 0
        root = Trie()
        root.add(words)
        return root.prefixCount(pref)

File: 2185-counting-words-with-a-given-prefix/2185-counting-words-with-a-given-prefix.cpp
class TrieNode{
    public:
    vector<TrieNode*> children;
    int count;

    TrieNode(){
        this->children.resize(26,nullptr);
        this->count = 0;
    }
};

class Trie{
    public:
    TrieNode* root;

    Trie(){
        root = new TrieNode();
    }

    void add(vector<string>& words){
        for(const string word : words){
            TrieNode* curr = this->root;
            for(const char c : word){
                if(!curr->children[c - 'a']) 
                curr->children[c - 'a'] = new TrieNode();
                curr = curr->children[c - 'a'];
                ++curr->count;
            }
        }
    }

    int getPrefixCount(string& pref){
        TrieNode* curr = this->root;
        for(const char c : pref){
            if(!curr->children[c - 'a']) return 0;
            curr = curr->children[c - 'a'];
        }
        return curr->count;
    }

    ~Trie(){
        delete root;
    }
};

class Solution {
public:
    int prefixCount(vector<string>& words, string pref) {
        int res = 0;
        Trie* root = new Trie();
        root->add(words);
        int count = root->getPrefixCount(pref);
        delete root;
        return count;
    }
};